### PR TITLE
chore: use cp vl when circuit is CRT based

### PR DESCRIPTION
**Pull request by jfrery**
closes https://github.com/zama-ai/concrete-ml-internal/issues/4393

### CODE REVIEW

**Review by andrei-stoian-zama — APPROVED** [review]
Looks good, just a minor doubt on whether the compiled model will always be compilable and CRT with n_bits=4. But otherwise great!

File: tests/quantization/test_quantized_module.py [hunk 1]
```diff
@@ -512,3 +512,49 @@ def test_quantized_module_initialization_error():
             ordered_module_output_names=None,  # This makes the combination invalid
             quant_layers_dict={"layer1": (["input1"], "QuantizedOp")},
         )
+
+
+@pytest.mark.parametrize("model_class, input_shape", [pytest.param(FC, (100, 32 * 32 * 3))])
+def test_crt_simulation_using_cp_vl(model_class, input_shape, default_configuration):
```
Comment (RomanBredehoft) [inline tests/quantization/test_quantized_module.py:518]: the test name is more about checking that rounding makes the circuit non-crt, there's nothing about simulation in this test, so maybe you can rename `test_crt_simulation_using_cp_vl` to `test_crt_to_non_crt_using_rouding` or else

File: src/concrete/ml/sklearn/base.py [hunk 2]
```diff
@@ -645,17 +645,19 @@ def predict(self, X: Data, fhe: Union[FheMode, str] = FheMode.DISABLE) -> numpy.
...
+                    # and the circuit encoding (CRT encoding should use Concrete VL)
+                    if USE_OLD_VL or is_crt_encoding:
                         predict_method = partial(
                             self.fhe_circuit.graph, p_error=self.fhe_circuit.p_error
                         )  # pragma: no cover
 
                     # Else, use the official simulation method
                     else:
                         predict_method = self.fhe_circuit.simulate
-
```
Comment (RomanBredehoft) [inline src/concrete/ml/sklearn/base.py:658]: better to keep a space between a code line and a following comment

File: src/concrete/ml/quantization/quantized_module.py [hunk 3]
```diff
@@ -522,14 +522,15 @@ def _fhe_forward(
...
-                if USE_OLD_VL:
+                # Whether the old simulation method should be used based on USE_OLD_VL
+                # FIXME: https://github.com/zama-ai/concrete-ml-internal/issues/4391
+                # and the circuit encoding (CRT encoding should use Concrete VL)
+                if USE_OLD_VL or is_crt_encoding:
                     predict_method = partial(
                         self.fhe_circuit.graph, p_error=self.fhe_circuit.p_error
                     )  # pragma: no cover
-
-                # Else, use the official simulation method
```
Comment (RomanBredehoft) [inline src/concrete/ml/quantization/quantized_module.py:532]: not sure why the comment was removed, and same, better to keep an empty line above !

File: src/concrete/ml/sklearn/base.py [hunk 4]
```diff
@@ -645,17 +645,19 @@ def predict(self, X: Data, fhe: Union[FheMode, str] = FheMode.DISABLE) -> numpy.
 
                 # If the inference should be executed using simulation
                 if fhe == "simulate":
+                    is_crt_encoding = self.fhe_circuit.statistics["packing_key_switch_count"] != 0
 
-                    # If the old simulation method should be used
-                    if USE_OLD_VL:
+                    # Whether the old simulation method should be used based on USE_OLD_VL
```
Comment (RomanBredehoft) [inline src/concrete/ml/sklearn/base.py:650]: I would rephrase this with : 
```
# If the virtual library method should be used 
# For now, use the virtual library when simulating circuits that uses CRT 
# decomposition because the official simulation is too slow 
# FIXME: https://github.com/zama-ai/concrete-ml-internal/issues/4391
```

File: src/concrete/ml/quantization/quantized_module.py [hunk 5]
```diff
@@ -522,14 +522,15 @@ def _fhe_forward(
 
             # If the inference should be executed using simulation
             if simulate:
+                is_crt_encoding = self.fhe_circuit.statistics["packing_key_switch_count"] != 0
 
-                # If the old simulation method should be used
-                if USE_OLD_VL:
+                # Whether the old simulation method should be used based on USE_OLD_VL
```
Comment (RomanBredehoft) [inline src/concrete/ml/quantization/quantized_module.py:527]: I would rephrase this with : 
```
# If the virtual library method should be used 
# For now, use the virtual library when simulating circuits that uses CRT 
# decomposition because the official simulation is too slow 
# FIXME: https://github.com/zama-ai/concrete-ml-internal/issues/4391
```

File: tests/quantization/test_quantized_module.py [hunk 6]
```diff
@@ -512,3 +512,49 @@ def test_quantized_module_initialization_error():
...
+    "Test simulation based on whether the circuit uses CRT encoding."
```
Comment (RomanBredehoft) [inline tests/quantization/test_quantized_module.py:519]: same here, maybe put "Tes that using rounding makes circuits non-CRT" or something like that

**Review by RomanBredehoft — CHANGES_REQUESTED** [review]
thanks for this update, I would just make some things a bit more clear !

File: src/concrete/ml/sklearn/base.py (hunk 4 above)
Comment (jfrery) [reply src/concrete/ml/sklearn/base.py:650]: in what?

File: src/concrete/ml/quantization/quantized_module.py (hunk 5 above)
Comment (jfrery) [reply src/concrete/ml/quantization/quantized_module.py:527]: I don't like this `# If the virtual library method should be used `. Feels like something else is coming to complete the sentence.

File: src/concrete/ml/sklearn/base.py (hunk 4 above)
Comment (RomanBredehoft) [reply src/concrete/ml/sklearn/base.py:650]: in this : https://github.com/zama-ai/concrete-ml/pull/619/files#r1568710561

**Comment by andrei-stoian-zama** [comment]
This also closes https://github.com/zama-ai/concrete-ml-internal/issues/4391 right ?

File: src/concrete/ml/quantization/quantized_module.py (hunk 3 above)
Comment (jfrery) [reply src/concrete/ml/quantization/quantized_module.py:532]: Added it back

File: src/concrete/ml/sklearn/base.py (hunk 2 above)
Comment (jfrery) [reply src/concrete/ml/sklearn/base.py:658]: Fixed

File: src/concrete/ml/quantization/quantized_module.py (hunk 5 above)
Comment (jfrery) [reply src/concrete/ml/quantization/quantized_module.py:527]: Anyway back to the original comment with some of your suggestion.

**Review by RomanBredehoft — APPROVED** [review]
great thanks !

**Comment by RomanBredehoft** [comment]
no @andrei-stoian-zama because the issue you ink is used as a fixme, until this is somehow fixed by concrete
